qip-0015.md

```diff
@@ -0,0 +1,59 @@
+
+### UTXO Key Organization
+Because we must trim blocks at depth, we must store the UTXOs created by each block on disk. The UTXO set is a flat database, not keyed by block hash, but rather keyed by transaction hash and index. One way to do this is to add the block number and denomination into the key as a prefix, and then iterate through all keys on disk with the block number and denomination as the prefix and trim the UTXOs under that prefix.
+The issue with this approach is that it requires the user or sender of a transaction to provide the block number and denomination into the input so the UTXO can be looked up for verification. Denomination is easy, as the user knows how much they will spend, but block number is slightly more difficult, as the block height that created a UTXO can change at the tip. A wallet might mistakenly mark that a UTXO was created in a certain block, only for that information to be invalidated 5 minutes later.
+Additionally, this approach increases the size of every input by 9 bytes - 8 bytes for the block height and 1 byte for the denomination, which increases storage requirements.
+
+A different approach is to store the 9 bytes in a different way - store the key of every UTXO created by a block on disk, but shorten the key to 9 bytes: The first 8 bytes of the txhash and the 1-byte denomination. This is a separate index that maps block height to shortened UTXO keys. The storage requirements are similar to the prior approach, but the user does not have to provide extra information that takes up bandwidth in transaction propagation. The UTXO index is not necessary, as all indices of a transaction for a certain block and denomination will be trimmed. In addition, only keys for UTXOs of denominations 0 through 6 need to be stored, as all other UTXOs will never be trimmed.
+
+The birthday problem formula provides an estimate of the entropy provided by an 8-byte hash. For 5 billion random elements, the chance of one unintentional collision is 50%. However, a malicious actor could intentionally create collisions. To account for this, I suggest a simple rule: if there is a duplicate key, and the denomination is the same, we delete both UTXOs. While this is incredibly unlikely, it would be very difficult to determine which UTXO was created in the block we are trying to trim without having the full historical block data, which most nodes will prune out.
+Even in such an unlikely case, the value that is trimmed will be below 1 Qi, so it is a reasonable approach to take to reduce data storage across all nodes in the network. If a malicious actor intentionally creates more duplicate keys than can reasonably be trimmed in a single block, the duplicate key is stored to disk, and the next block will continue the trimming process to remove the rest of the duplicates.
```

Comment:
> If a malicious actor intentionally creates more duplicate keys than can reasonably be trimmed in a single block, the duplicate key is stored to disk, and the next block will continue the trimming process to remove the rest of the duplicates.

How do you determine what "can reasonably be done in a single block"? Either it is part of the block processing rules or it isnt.

Reply:
Added a suggested limit per key per block